src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,288 @@
+
+"""
+ceph-mgr orchestrator interface
+
+This is a DRAFT for discussion.
+
+Goal: enable UI workflows for cluster service management
+      (such as creating OSDs, in addition to stateless services)
+      using common concepts that are implemented by
+      diverse backends such as Rook, DeepSea, ceph-ansible
+
+Concepts:
+    "Stateful service": a daemon that uses local storage, such as OSD or mon.
+    "Stateless service": a daemon that doesn't use any local storage, such
+                         as an MDS, RGW, nfs-ganesha, iSCSI gateway.
+    "Label": arbitrary string tags that may be applied by administrators
+             to nodes.  Typically administrators use labels to indicate
+             which nodes should run which kinds of service.  Labels are
+             advisory (from human input) and do not guarantee that nodes
+             have particular physical capabilities.
+
+Constraints:
+    1. The orchestrator is to be the source of truth for
+       all the physical information, and will be queried directly
+       as needed (i.e. no in-Ceph database of hardware etc).
+    2. The orchestrator handles placement of collections of stateless
+       services.
+    3. The orchestrator accepts explicit placement of individual stateful
+       services, and optionally also accepts label-based automatic placement.
+       (i.e. it *must* support "create OSD at host1:/dev/sdb", and it *may*
+        support "create OSDs on nodes with label=ceph-osd")
```

Comment:
how about node name regex?

Reply:
Does k8s support targeting hosts by regex?  I would have thought the general answer to regex host selection is "that's what labels are for"

Reply:
As another thought, even if k8s doesn't support targeting hosts by regex, the k8s orchestrator python code could take in a regex and do the necessary work to filter k8s nodes itself.

I think accepting node's with a regex makes a lot of sense. Even just shell-style globbing like `ceph-*` would be useful.

Reply:
No argument about it being useful, but supporting it means moving all the host filtering logic up from the orchestrator into ceph-mgr, which prevents orchestrators from dynamically rescheduling services (see later comment).